### PR TITLE
Moving Axes to Redux

### PR DESCRIPTION
**Pull request by PavelVanecek**
## Description

One step closer in my quest of rendering Tooltip from Redux store alone.

Tooltip is not yet functional - for that it needs Tooltip.shared and Scatter - but it's almost there. Builds are green so I figured we might as well merge it.

## Related Issue

https://github.com/recharts/recharts/discussions/3717

## Motivation and Context

No direct DOM access

## How Has This Been Tested?

npm test

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] I have added a storybook story or extended an existing story to show my changes

### CODE REVIEW

**Review by ckifer — APPROVED** [review]

**Comment by ckifer** [comment]
Nice

File: src/state/selectors.ts [hunk 1]
```diff
@@ -26,29 +31,60 @@ function getSliced<T>(arr: ReadonlyArray<T>, startIndex: number, endIndex: numbe
...
+      return undefined;
+    }
+    const { chartData, dataStartIndex, dataEndIndex } = chartDataState;
+
+    /*
+     * If a payload has data specified directly from the graphical item, prefer that.
+     * Otherwise, fill in data from the chart level, using the same index.
+     */
+    return tooltipItemPayloads.map(({ dataDefinedOnItem, settings }) => {
+      const finalData = dataDefinedOnItem ?? chartData;
```
Comment (ckifer) [inline src/state/selectors.ts:72]: this will need changed to a check for length or `dataDefinedOnItem` will need to be null/undefined

Comment (PavelVanecek) [reply src/state/selectors.ts:72]: Possible. Do we have all the cases covered with tests?

Comment (ckifer) [reply src/state/selectors.ts:72]: no https://github.com/recharts/recharts/blob/3.x/src/cartesian/Scatter.tsx#L205
😅

Comment (ckifer) [reply src/state/selectors.ts:72]: https://github.com/search?q=repo%3Arecharts%2Frecharts%20data%3A%20%5B%5D&type=code

Those first three there will all return `dataDefinedOnItem` every time

Comment (PavelVanecek) [reply src/state/selectors.ts:72]: That might actually explain some of the failures I saw.

File: src/state/selectors.ts [hunk 2]
```diff
@@ -26,29 +31,60 @@ function getSliced<T>(arr: ReadonlyArray<T>, startIndex: number, endIndex: numbe
...
+      if (tooltipAxis?.dataKey && !tooltipAxis?.allowDuplicatedCategory) {
+        tooltipPayload = findEntryInArray(sliced, tooltipAxis.dataKey, activeLabel);
+      } else {
+        tooltipPayload = sliced?.[activeIndex];
+      }
+      return getTooltipEntry({
+        tooltipEntrySettings: settings,
+        dataKey: settings?.dataKey,
+        payload: tooltipPayload,
+        value: getValueByDataKey(tooltipPayload, settings?.dataKey),
```
Comment (ckifer) [inline src/state/selectors.ts:86]: I think this also needs to be `settings?.dataKey ?? tooltipAxis?.dataKey`

Else no value will be rendered in tooltip when the axis has the dataKey (at least for Scatter)

Comment (ckifer) [reply src/state/selectors.ts:86]: Also really wonder how Cell is going to work :/

File: src/state/selectors.ts [hunk 3]
```diff
@@ -26,29 +31,60 @@ function getSliced<T>(arr: ReadonlyArray<T>, startIndex: number, endIndex: numbe
...
+  selectTooltipAxis,
+  selectActiveLabel,
+  (
+    tooltipState: TooltipState,
+    chartDataState: ChartDataState,
+    tooltipAxis: BaseAxisProps | undefined,
+    activeLabel: string | undefined,
+  ) => {
+    const { activeIndex, tooltipItemPayloads } = tooltipState;
+    if (activeIndex === -1) {
```
Comment (ckifer) [inline src/state/selectors.ts:62]: activeShape isn't getting reset anymore
https://63da8268a0da9970db6992aa-pkiagidnpj.chromatic.com/?path=/story/examples-scatterchart--simple

https://63da8268a0da9970db6992aa-pkiagidnpj.chromatic.com/?path=/docs/api-cartesian-bar--docs

Hover over those

Comment (PavelVanecek) [reply src/state/selectors.ts:62]: It gets reset when I hover on another element, and it does not reset (but it should) on mouse leave, is that correct?

File: src/state/selectors.ts (hunk 1 above)
Comment (ckifer) [reply src/state/selectors.ts:72]: `Scatter` was returning nothing for Tooltip data. Now it is (once the dataKey issue is good to go too)

We also fix a bug where Scatter tooltips never get color styles when we do this

File: src/state/selectors.ts (hunk 3 above)
Comment (ckifer) [reply src/state/selectors.ts:62]: correct yep

Comment (PavelVanecek) [reply src/state/selectors.ts:62]: I will have a look

File: src/state/selectors.ts (hunk 1 above)
Comment (PavelVanecek) [reply src/state/selectors.ts:72]: And what's worse, the `getComposedData` will re-inject the `data` prop and overwrite whatever it was that was there before so we lose access to it.

Soooo a new blocker to this blocker, we need to remove `getComposedData` first.

File: src/state/selectors.ts (hunk 3 above)
Comment (PavelVanecek) [reply src/state/selectors.ts:62]: https://github.com/recharts/recharts/pull/4527

File: src/state/selectors.ts (hunk 2 above)
Comment (PavelVanecek) [reply src/state/selectors.ts:86]: Also this is, in the tooltipAxis case, applying the dataKey twice? This looks like it shouldn't work at all

**Pull request merged**